### PR TITLE
Freemabd1/adding cohort builder

### PR DESCRIPTION
**Pull request by freemabd1**
Initial commit of the cohort builder. Will have to run npm install inside the ui directory

### CODE REVIEW

File: ui/src/app/views/cohort-builder/search/cohort-builder/cohort-builder.component.ts [hunk 1]
```diff
@@ -0,0 +1,205 @@
+import { Component, OnInit, ChangeDetectorRef, OnDestroy, Inject } from '@angular/core';
```
Comment (blrubenstein) [inline ui/src/app/views/cohort-builder/search/cohort-builder/cohort-builder.component.ts]: Remove extra spacing from imports.

File: ui/src/app/app-routing.module.ts [hunk 2]
```diff
@@ -4,14 +4,14 @@
...
-import {CohortBuilderComponent} from 'app/views/cohort-builder/component';
-import {CohortBuilderPlaceholderComponent} from 'app/views/cohort-builder-placeholder/component';
+import {CohortBuilderComponent} from 'app/views/cohort-builder/search/cohort-builder/cohort-builder.component';
 import {CohortEditComponent} from 'app/views/cohort-edit/component';
 import {HomePageComponent} from 'app/views/home-page/component';
 import {LoginComponent} from 'app/views/login/component';
 import {WorkspaceComponent} from 'app/views/workspace/component';
 import {WorkspaceEditComponent} from 'app/views/workspace-edit/component';
 import {SelectRepositoryComponent} from 'app/views/select-repository/component';
+import { CohortReviewComponent } from 'app/views/cohort-builder/review/cohort-review/cohort-review.component';
```
Comment (jscherdin) [inline ui/src/app/app-routing.module.ts]: That's what the CLI does out of the box

Comment (blrubenstein) [inline ui/src/app/app-routing.module.ts]: I think its fine if we do want to have the spaces, the big thing for me is consistency. So if we are going to do the spaces we should add them to our other imports.

Comment (jscherdin) [inline ui/src/app/app-routing.module.ts]: That's my whole point. You are saying we should change something but no decision has ever been made on this. These are things we all need to talk about and decide. And in this case, no spaces breaks the default CLI behavior.

Comment (blrubenstein) [inline ui/src/app/app-routing.module.ts]: Sure, sorry I didn't make the consistency thing clear. That's my bad.

File: ui/src/app/views/cohort-builder/search/cohort-builder/cohort-builder.component.ts [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+              private broadcastService: BroadcastService,
+              private router: Router,
+              @Inject(DOCUMENT) private document: any) {}
+
+  ngOnInit() {
+    this.searchGroups = [new SearchGroup()];
+    this.exclusionGroups = [new SearchGroup('Exclude participants where:')];
+
+    this.broadcastService.updatedCounts$
+      .takeWhile(() => this.alive)
```
Comment (jmandel) [inline ui/src/app/views/cohort-builder/search/cohort-builder/cohort-builder.component.ts]: This `takeWhile` might be removed, in favor of unsubscribing via `ngOnDestroy`, unless I'm missing something? The subscribe function returns an unsubscribe function that can be stored and called when the time comes, which avoids maintaining this "alive" state. @freemabd1 not sure if this works for you, but it's worth knowing about if you haven't seen it.

File: ui/src/app/views/cohort-builder/search/cohort-builder/cohort-builder.component.ts [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+        this.updateGroupSetRemovedSearchResult(change.searchGroup);
+        this.updateTotalSet();
+        this.updateCharts();
+      });
+  }
+
+  addSearchGroup() {
+    this.searchGroups.push(new SearchGroup());
+    this.changeDetectorRef.detectChanges();
+    const scrollableDiv = window.document.getElementById('scrollable-groups');
```
Comment (jmandel) [inline ui/src/app/views/cohort-builder/search/cohort-builder/cohort-builder.component.ts]: Is it possible to use https://stackoverflow.com/a/35209681 or a similar technique to get a hold of this element @freemabd1 ?

File: ui/src/app/views/cohort-builder/search/cohort-builder/cohort-builder.component.ts [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+        set = intersection(set,
+            group.groupSet,
+            (object) => object.val);
+      }
+    });
+    return set;
+  }
+
+  private updateCharts() {
+    let female = 0;
```
Comment (jmandel) [inline ui/src/app/views/cohort-builder/search/cohort-builder/cohort-builder.component.ts]: For this kind of totaling logic @freemabd1, consider a `reduce` like

```typescript
    const summary = this.totalSet.reduce( (totals, item, index) => {
      let [gender, race] = item.val.split(',');
      gender = gender || 'Unknown.gender';
      race = race || 'Unknown.race';
      return {...totals,
        [gender]: (totals[gender] || 0) + 1,
        [race]: (totals[race] || 0) + 1,
      };
    }, { });
    const genderData = [
      ['Gender', 'Count', { role: 'style' }],
      ['Female', summary['F'], 'blue'],
      ['Male', summary['M'], 'red'],
      ['Unknown', summary['Unknown.gender'], 'gray']];
    const raceData = [
      ['Race', 'Count Per'],
      ['African American', summary['B']],
      ['Asian/Pacific', summary['A']],
      ['Caucasian', summary['W']],
      ['Native American', summary['I']],
      ['Hispanic', summary['H']],
      ['Other', summary['N']],
      ['Unknown', summary['Unknown.race']]];
```

File: ui/src/app/views/cohort-builder/search/cohort-builder/cohort-builder.component.ts (hunk 3 above)
Comment (freemabd1) [inline ui/src/app/views/cohort-builder/search/cohort-builder/cohort-builder.component.ts]: I'm not attached to takeWhile, so whatever everyone thinks is the best approach is fine with me.

File: ui/src/app/views/cohort-builder/search/cohort-builder/cohort-builder.component.ts (hunk 4 above)
Comment (freemabd1) [inline ui/src/app/views/cohort-builder/search/cohort-builder/cohort-builder.component.ts]: nice.. i will give this a try.

File: ui/src/app/views/cohort-builder/search/cohort-builder/cohort-builder.component.ts (hunk 5 above)
Comment (freemabd1) [inline ui/src/app/views/cohort-builder/search/cohort-builder/cohort-builder.component.ts]: Thanks for the tip Josh this is very helpful.

File: ui/src/app/views/cohort-builder/search/cohort-builder/cohort-builder.component.ts (hunk 3 above)
Comment (jmandel) [inline ui/src/app/views/cohort-builder/search/cohort-builder/cohort-builder.component.ts]: It's good practice to make sure we destroy any subscriptions as soon as they are no longer needed, so I'd prefer unsubscribe Z a general pattern

**Pull request merged**